### ISSUE TITLE
Compile error on Windows 64-bit with MinGW

### ISSUE THREAD

**Issue by nickhardcastle** [CERN/TIGRE#13]
Hi,

Windows 7 64-bit running Matlab 2015b and CUDA 7.5 - I've installed MinGW compiler and can compile other mex files ok in matlab. When I run Compile.m for TIGRE I get the following compilation error:

---

Compiling TIGRE source...
This may take a couple of minutes....
Error using mex
No supported compiler or SDK was found. You can install the freely available MinGW-w64 C/C++ compiler; see Install MinGW-w64 Compiler. For more options, see  http://www.mathworks.com/support/compilers/R2015b/win64.html.

Error in Compile (line 37)
        mex -largeArrayDims ./Source/Ax.cpp ./Source/ray_interpolated_projection.cu ./Source/Siddon_projection.cu
        ./Source/ray_interpolated_projection_parallel.cu ./Source/Siddon_projection_parallel.cu -outdir
##         ./Mex_files/win64

Any suggestion why mex is specifically not working for TIGRE? I'm assuming (hoping) it's something simple I've not done.

Cheers,

Nick

**Comment by chappjc** [comment]
This is just MATLAB setup of C++ compiler.  You can do as suggested and download and install MinGW or you can use Visual Studio.  Then run `mex -setup C` and `mex -setup C++`

**Comment by nickhardcastle** [comment]
Thanks - I went through the above steps as described in the installation instructions as the first step after installing MinGW and had no errors. Running the mex command on the matlab examples also has no errors. It's only when I try and run the mex command on TIGRE files do I get the above error.

**Comment by AnderBiguri** [comment]
Hi,

If you are completely sure about the correct installation of your compiler,
then it could be the mex_CUDA_win64.xml file.  I have never modified that
file  for compiler, but it definitely has some visual studio environment
paths on it.

You can find that file in your PC (some modifications on the CUDA part were
made for TIGRE) on `matlabroot\toolbox\distcomp\gpu\extern\src\mex\win64`

Unfortunately I lack the knowledge to help you more here. I am currently
working on win7x64 with MATLAB2014b with Visual Studio compiler and it does
work fine. I guess you can always install VS as a last option but it would
be nice to make it work for any cases.

Ander

**Comment by AnderBiguri** [comment]
When I said Visual Studio (VS) I meant "Microsoft Windows SDK 7.1 ,
requires .NET Framework 4.0" , sorry.
Ander

**Comment by nickhardcastle** [comment]
OK - what about the mex_CUDA_win64.xml that comes with TIGRE? I don't have the distributed computing toolbox so that directory doesn't exist. Is this toolbox needed? I can't see the system/matlab requirements documented...

If I have no success, and the distributed computing toolbox isn't required, I'll give the SDK 7.1 a go as suggested.

Thanks!

**Comment by AnderBiguri** [comment]
Hi,

Yes apologies,  TIGRE should work without any toolbox no problem, I was
just suggesting a work around.  I am not even sure if the xml is the
problem,   I just haven't tested it with minGW.

I will try to look further into this,  but indeed,  sdk 7.1 will almost
certainly work.

**Comment by AnderBiguri** [comment]
Hi again,
First, just in case, make sure you have MinGW 4.9.2 . It has to be this version.

Looking around I found out that there are still some problems with minGW even if it is supported. You will definitely have memory leaks when `mexErrMsgIdAndTxt ` is used (every error in TIGRE uses that function) and MATLAB will crash if any exception happens in the mex files. [See here for more information](https://uk.mathworks.com/help/matlab/matlab_external/install-mingw-support-package.html).

Unfortunately, with the current format of TIGRE, MinGW will create mex files that will memory leak and crash MATLAB. I may look into fixing this, but not sure how much is worth it, as it is known that MATLABs compatibility for compilers and mex stuff changes all the time, version by version. 

I will leave the Issue open, and feel free to try to modify the cpp/cu files for compatibility (if you get MinGW to work) and push a commit, but unfortunately I personally do not have the time to look at this now (sorry), as it would require looking problems in different MATLAB versions etc. 

Long story short: Use SDK7.1 as MinGW will not work, even if you get it to compiler.

Best,
Ander

**Comment by fribert** [comment]
I'm getting the same error message with SDK 7.1. , .NET framework 4.6, Matlab 2016a (and b) and Windows 10 64-bit. I was originally using MinGW, read this and changed my compiler to SDK 7.1. Still keep getting the same error.

>Error using mex
>No supported compiler or SDK was found. For options, visit
>http://www.mathworks.com/support/compilers/R2016a/win64.html.
>Error in Compile (line 37)
>       mex -largeArrayDims ./Source/Ax.cpp ./Source/ray_interpolated_projection.cu
>        ./Source/Siddon_projection.cu ./Source/ray_interpolated_projection_parallel.cu
>        ./Source/Siddon_projection_parallel.cu -outdir ./Mex_files/win64 

All other mex files work normally. Have you tried whether the Toolbox works with Windows 10?

**Comment by AnderBiguri** [comment]
Hi @fribert ,
Yes, my colleague has TIGRE working on a win10 laptop without any problem.
However, TIGRE does not give that error message, MATLAB does, so it must me something on the MATLAB side.  When you say "All other mex files work normally" you mean mex files outside TIGRE or the other TIGRE mex files? 

Ander

**Comment by AnderBiguri** [comment]
Hi @fribert 
MATLAB says NET Framerwork 4.0, not 4.6. I'd try to tick to what EXACTLY MATLAB suggests. They are known for not having version compatibility.
Ander

**Comment by dewan85** [comment]
hello
I am using windows 7 64 bit, Matlab 2015 a, SDK 7.1 and netframework 4.0. But i am not able to compile the code. Please help me. I am attaching the error below

error :+1: 
>> Compile
MEX configured to use 'Microsoft Visual C++ 2008 (C)' for C language compilation.
Warning: The MATLAB C and Fortran API has changed to support MATLAB
	 variables with more than 2^32-1 elements. In the near future
	 you will be required to update your code to utilize the
	 new API. You can find more information about this at:
	 http://www.mathworks.com/help/matlab/matlab_external/upgrading-mex-files-to-use-64-bit-api.html.

To choose a different C compiler, select one from the following:
Microsoft Visual C++ 2008 (C)  mex -setup:C:\Users\dewan\AppData\Roaming\MathWorks\MATLAB\R2015a\mex_C_win64.xml C
Microsoft Windows SDK 7.1 (C)  mex -setup:'C:\Program Files\MATLAB\MATLAB Production Server\R2015a\bin\win64\mexopts\winsdk-7.1_c.xml' C

To choose a different language, select one from the following:
 mex -setup C++ 
 mex -setup FORTRAN
Compiling TIGRE source...
This may take a couple of minutes....
Error using mex
No supported compiler or SDK was found. For options, visit  http://www.mathworks.com/support/compilers/R2015a/win64.html.

Error in Compile (line 37)
        mex -largeArrayDims ./Source/Ax.cpp ./Source/ray_interpolated_projection.cu ./Source/Siddon_projection.cu
        ./Source/ray_interpolated_projection_parallel.cu ./Source/Siddon_projection_parallel.cu -outdir ./Mex_files/win64

**Comment by AnderBiguri** [comment]
Hi,
Have you tried running mex -setup C++?
Ander

On 3 Nov 2016 1:10 p.m., "dewan85" <notifications@github.com> wrote:

hello
I am using windows 7 64 bit, Matlab 2015 a, SDK 7.1 and netframework 4.0.
But i am not able to compile the code. Please help me. I am attaching the
error below

error 👍

Compile
MEX configured to use 'Microsoft Visual C++ 2008 (C)' for C language
compilation.
Warning: The MATLAB C and Fortran API has changed to support MATLAB
variables with more than 2^32-1 elements. In the near future
you will be required to update your code to utilize the
new API. You can find more information about this at:
http://www.mathworks.com/help/matlab/matlab_external/
upgrading-mex-files-to-use-64-bit-api.html.

To choose a different C compiler, select one from the following:
Microsoft Visual C++ 2008 (C) mex -setup:C:\Users\dewan\AppData\
Roaming\MathWorks\MATLAB\R2015a\mex_C_win64.xml C
Microsoft Windows SDK 7.1 (C) mex -setup:'C:\Program Files\MATLAB\MATLAB
Production Server\R2015a\bin\win64\mexopts\winsdk-7.1_c.xml' C

To choose a different language, select one from the following:
mex -setup C++
mex -setup FORTRAN
Compiling TIGRE source...
This may take a couple of minutes....
Error using mex
No supported compiler or SDK was found. For options, visit
http://www.mathworks.com/support/compilers/R2015a/win64.html.

Error in Compile (line 37)
mex -largeArrayDims ./Source/Ax.cpp ./Source/ray_interpolated_projection.cu
./Source/Siddon_projection.cu
./Source/ray_interpolated_projection_parallel.cu ./Source/Siddon_projection_
parallel.cu -outdir ./Mex_files/win64

—

You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/CERN/TIGRE/issues/13#issuecomment-258137618>, or mute
the thread

**Comment by dewan85** [comment]
hello

I have tried mex -setup C++. But still i receive the same error.

dewan

**Comment by AnderBiguri** [comment]
What happens when you run mex - setup C++?  Does matlab complain?

**Comment by dewan85** [comment]
i think there is no complain from matlab. i am attaching below.

MEX configured to use 'Microsoft Windows SDK 7.1 (C++)' for C++ language compilation.
Warning: The MATLAB C and Fortran API has changed to support MATLAB
	 variables with more than 2^32-1 elements. In the near future
	 you will be required to update your code to utilize the
	 new API. You can find more information about this at:
	 http://www.mathworks.com/help/matlab/matlab_external/upgrading-mex-files-to-use-64-bit-api.html.

To choose a different C++ compiler, select one from the following:
Microsoft Visual C++ 2008  mex -setup:'C:\Program Files\MATLAB\MATLAB Production Server\R2015a\bin\win64\mexopts\msvcpp2008.xml' C++
Microsoft Windows SDK 7.1 (C++)  mex -setup:C:\Users\dewan\AppData\Roaming\MathWorks\MATLAB\R2015a\mex_C++_win64.xml C++

**Comment by AnderBiguri** [comment]
@dewan85 Can you please try to compile something else that is not TIGRE? sine if the MATLAB examples. This way we can see if the problem is on TIGRE or on your local configuration

**Comment by martijn-d** [comment]
Hi all,

I've had similar issues. Here are some pointers to make the compilation work:
- Make sure that your CUDA Toolkit is compatible with Matlab (see www.mathworks.com/help/distcomp/release-notes.html)
- Make sure your CUDA Toolkit is compatible with your Visual Studio Compiler
- The CUDA Toolkit installer overrides your current graphical driver, which makes it unrecognizable to Matlab.  Make sure your NVIDIA driver is up-to-date and install the toolkit without the graphic driver option. A correct installation can be checked with the 'gpuDevice' command.

Cheers,
Martijn

**Comment by emrtpl** [comment]
Hello all, unfortunately i have also experiencing the same error with SDK 7.1 compiler. I checked nvidia driver and cuda toolkit but they are working fine. Is there any possible solutions i can try to overcome this error? I desperately need to see fdk reconstructed version of my projection data to compare with my own algorithm. I already modified Kyungsang Kim's matlab program to working on my own data but the only thing i get was a slice of reconstructed images with full of meaningless circles.

**Comment by AnderBiguri** [comment]
Hi @emrtpl,
Can I get more information about what error you are getting? Can I also get more information about you rMATLAB version, CUDA version and SDK version? What happens when you do mex -setup? 
I see that you are not using MinGW, so I propose we open a new Issue (or send me an email) instead of adding it to this one.
Best
Ander

**Comment by emrtpl** [comment]
Hi @AnderBiguri 

I am using matlab r2016a, cuda 8 and sdk 7.1. When i tried to run compile.m file, i got message below:

```
>> Compile
MEX configured to use 'Microsoft Windows SDK 7.1 (C)' for C language compilation.
Warning: The MATLAB C and Fortran API has changed to support MATLAB
	 variables with more than 2^32-1 elements. In the near future
	 you will be required to update your code to utilize the
	 new API. You can find more information about this at:
	 http://www.mathworks.com/help/matlab/matlab_external/upgrading-mex-files-to-use-64-bit-api.html.

To choose a different language, select one from the following:
 mex -setup C++ 
 mex -setup FORTRAN
Compiling TIGRE source...
This may take a couple of minutes....
Error using mex
No supported compiler or SDK was found. For options, visit http://www.mathworks.com/support/compilers/R2016a/win64.html.

Error in Compile (line 37)
        mex -largeArrayDims ./Source/Ax.cpp ./Source/ray_interpolated_projection.cu ./Source/Siddon_projection.cu
        ./Source/ray_interpolated_projection_parallel.cu ./Source/Siddon_projection_parallel.cu -outdir ./Mex_files/win64
```
 

output of mex-setup

```
>> mex -setup
MEX configured to use 'Microsoft Windows SDK 7.1 (C)' for C language compilation.
Warning: The MATLAB C and Fortran API has changed to support MATLAB
	 variables with more than 2^32-1 elements. In the near future
	 you will be required to update your code to utilize the
	 new API. You can find more information about this at:
	 http://www.mathworks.com/help/matlab/matlab_external/upgrading-mex-files-to-use-64-bit-api.html.

To choose a different language, select one from the following:
 mex -setup C++ 
 mex -setup FORTRAN
```

and details of gpudevice

```
>> gpuDevice()

ans = 

  CUDADevice with properties:

                      Name: 'Quadro K5000'
                     Index: 1
         ComputeCapability: '3.0'
            SupportsDouble: 1
             DriverVersion: 8
            ToolkitVersion: 7.5000
        MaxThreadsPerBlock: 1024
          MaxShmemPerBlock: 49152
        MaxThreadBlockSize: [1024 1024 64]
               MaxGridSize: [2.1475e+09 65535 65535]
                 SIMDWidth: 32
               TotalMemory: 4.2950e+09
           AvailableMemory: 4.1018e+09
       MultiprocessorCount: 8
              ClockRateKHz: 705500
               ComputeMode: 'Default'
      GPUOverlapsTransfers: 1
    KernelExecutionTimeout: 1
          CanMapHostMemory: 1
           DeviceSupported: 1
            DeviceSelected: 1


```

**Comment by AnderBiguri** [comment]
@emrtpl  Please open a new issue and we can continue this discussion there.

**Comment by rizwindu** [comment]
I have come across a similar issue but compiling on Linux. I believe the issue that, from MATLAB 2015b, the command mexcuda is used to compile the cuda files.

http://uk.mathworks.com/help/distcomp/mexcuda.html

I'm currently unable to test this, but hope to do so at some point soon.

Best wishes

Ben

**Comment by AnderBiguri** [comment]
Hi @rizwindu ,
Similar issue than which one?

Technically `mexcuda` is not needed at all to compile TIGRE. a proper `mex` call with the `nvcc` compiler liked should do the job. I have personally not tested TIGRE in linux unfortunately, meaning that maybe the `mex_CUDA_glnxa64.xml` file is not properly set up with the right paths. Additionally, in widows,  the compiler seems to require an enviromet variable for `CUDA_PATH` that can be set up with `setenv('CUDA_PATH',yourpath)`. Not tested in Linux, but I think it should be similar

**Comment by rizwindu** [comment]
The compile error I get is similar to this one pasted above (from memory, but pretty sure it is the same as this)

```
Error using mex
No supported compiler or SDK was found. For options, visit
http://www.mathworks.com/support/compilers/R2016a/win64.html.
Error in Compile (line 37)
mex -largeArrayDims ./Source/Ax.cpp ./Source/ray_interpolated_projection.cu
./Source/Siddon_projection.cu ./Source/ray_interpolated_projection_parallel.cu
./Source/Siddon_projection_parallel.cu -outdir ./Mex_files/win64
```

Setup is MATLAB 2016b on Ubuntu with the latest CUDA Toolkit 8.0. I wasn't able to test compiling some simple .cu files, but will do so when I get an opportunity.

Best wishes

Ben

**Comment by AnderBiguri** [comment]
@rizwindu  and @nickhardcastle 

When the error with `mex` appears there are 2 possible options. 

1. mex is not properly set up
2. The compiler does not now how to compile `.cu` files. The most likely scenario for this one is the lack of a proper link to `nvcc` compiler. Try `getenv('CUDA_PATH')`. This must give you the path of the compiler if properly set up. 

Ander

**Comment by rizwindu** [comment]
Hi Ander

A quick update on this, but probably not resolving any issues.

Installed the nvidia CUDA toolkit from the Ubuntu package manager I get access to nvcc.

I can confirm nvcc works by compiling a simple hello.cu file from the terminal (like this: https://www.pdc.kth.se/resources/computers/historical-computers/zorn/how-to/how-to-compile-and-run-a-simple-cuda-hello-world). It also runs at the terminal simply using `./hello.out`

In matlab,` getenv('CUDA_PATH')` returns `''`.

If I run Compile.m I get the following error:

```
Building using 'nvcc'.
Error using mex
nvcc fatal     : Unsupported gpu architecture 'compute_13'

Error in Compile (line 64)
```

However, I tried a simple test of mexcuda by running `mexcuda voxel_backprojection.cu` in the Source directory and this completes successfully. Perhaps confusingly, running `mex voxel_backprojection.cu` says that no supported compiler was found.

I edited mex_CUDA_glnxa64.xml to update the location of libcudart.o. If you install nvidia-toolkit from the Ubuntu repo this gets put in `/usr/lib/x86_64-linux-gnu`

I also edited Compile.m to use `mexcuda` on lines 64-72 and now it completes successfully.

**Comment by AnderBiguri** [comment]
Hi @rizwindu ,

To avoid that error just go to the XML file I mentioned before and remove `-gencode=arch=compute_13,code=sm_13` from line  34. The xml is set up to compile all architectures, but sometimes CUDA complains for very very old arch (such as 1.3). Your GPU is likely to be avobe 3. Removing that piece of text should result in compilation without errors.

**Comment by rizwindu** [comment]
Hi

Thanks for the tip to remove the unsupported architecture.

To confirm, it all seems to compile now if I use `mexcuda` in Compile.m

**Comment by AnderBiguri** [comment]
Thats good!
Can you send me the modified Compile file so I can try it myself?
It does compile, but does it run? Can you run demo 3 for example?
Ander

**Comment by rizwindu** [comment]
Attached is the updated Compile.m (although changed to .txt so it uploads in this comment field.

[Compile.txt](https://github.com/CERN/TIGRE/files/831440/Compile.txt)

I ran d03_generateData with no errors.

Best wishes

Ben

**Comment by AnderBiguri** [comment]
The only supported CUDA compiler on Windows is Visual  Studio, according to Mathworks. I am closing this issue.
